RZCardEntry/RZCardEntry/CardType.swift

```diff
@@ -0,0 +1,203 @@
+        }
+    }
+
+    // from: https://gist.github.com/cwagdev/635ce973e8e86da0403a
+    static func luhnCheck(_ cardNumber: String) -> Bool {
+        var sum = 0
+        let reversedCharacters = cardNumber.characters.reversed().map { String($0) }
+        for (idx, element) in reversedCharacters.enumerated() {
+            guard let digit = Int(element) else { return false }
+            switch ((idx % 2 == 1), digit) {
+            case (true, 9): sum += 9
+            case (true, 0...8): sum += (digit * 2) % 9
+            default: sum += digit
+            }
+        }
+        return sum % 10 == 0
+    }
+
+}
+
+//MARK: - CardState
+
+enum CardState {
+    case identified(CardType)
+    case indeterminate([CardType])
+    case invalid
+}
+
+extension CardState: Equatable {}
+func ==(lhs: CardState, rhs: CardState) -> Bool {
+    switch (lhs, rhs) {
+    case (.invalid, .invalid): return true
+    case (.indeterminate, .indeterminate): return true
+    case (let .identified(card1), let .identified(card2)): return card1 == card2
+    default: return false
+    }
+}
+
+extension CardState {
+
+    static func fromNumber(_ cardNumber: String) -> CardState {
+        for cardType in CardType.allValues {
+            if cardType.valid(cardNumber) {
+                return .identified(cardType)
+            }
+        }
+        return .invalid
+    }
+
+    static func fromPrefix(_ cardPrefix: String) -> CardState {
+        guard !cardPrefix.isEmpty else {
+            return .indeterminate(CardType.allValues)
+        }
+        
+        let possibleTypes = CardType.allValues.filter { $0.prefixValid(cardPrefix) }
+        guard let card = possibleTypes.first else {
+            return .invalid
+        }
+        if possibleTypes.count == 1 {
+            return .identified(card)
+        }
+        else {
+            return .indeterminate(possibleTypes)
+        }
+    }
+}
+
+//MARK: - PrefixContainable
+
+fileprivate protocol PrefixContainable {
+
+    func prefixMatches(_ text: String) -> Bool
+    
+}
+
+extension ClosedRange: PrefixContainable {
+
+    func prefixMatches(_ text: String) -> Bool {
+
+        //cannot include Where clause in protocol conformance, so have to ensure Bound == String :(
+        guard !text.isEmpty, let lower = lowerBound as? String, let upper = upperBound as? String else { return false }
+
+        let trimmedRange: ClosedRange<String> = {
+            let length = text.characters.count
+            let trimmedStart = String(lower.characters.prefix(length))
+            let trimmedEnd = String(upper.characters.prefix(length))
+            return trimmedStart...trimmedEnd
+        }()
+
+        let trimmedText = String(text.characters.prefix(trimmedRange.lowerBound.characters.count))
+        return trimmedRange ~= trimmedText
+    }
+
+}
+
+extension String: PrefixContainable {
+
+    func prefixMatches(_ text: String) -> Bool {
+        return hasPrefix(text) || text.hasPrefix(self)
```

Comment:
I agree that `"AA".hasPrefix("AAB") should be false. I'm trying to express `"AA".sharesPrefixWith("AAB")`

The protocol conformance is for the nice array of both `String`s and `ClosedRange`s.
Perhaps something like `sharesPrefix` may be more appropriate for the `String` extension, but I suppose I went with `prefixMatches` because of the pattern matching happening in the `ClosedRange` extension.

All told, I need to be able say something like `"4".prefixMatches(visaNumber)` and `"51"..."55".prefixMatches(masterCardNumber)`